### PR TITLE
Fatptr review

### PR DESCRIPTION
**Pull request by priyasiddharth**
This contains the complete implementation so far including shadow mem

All the tests pass, but the store_load test take some time to complete. Since the total runtime is 6s - it may be expected.

```
-- Testing: 6 tests, 6 threads --
PASS: Seahorn :: fat_ptr_get_set_test_unsat.ll (1 of 6)
PASS: Seahorn :: fat_ptr_simple_unsat.ll (2 of 6)
PASS: Seahorn :: fat_ptr_simple_sat.ll (3 of 6)
PASS: Seahorn :: fat_ptr_outofbounds_01_sat.ll (4 of 6)
PASS: Seahorn :: fat_ptr_outofbounds_02_unsat.ll (5 of 6)
PASS: Seahorn :: fat_ptr_store_load_test_unsat.ll (6 of 6)
Testing Time: 6.74s
  Expected Passes    : 6
```

### CODE REVIEW

**Comment by agurfinkel** [comment]
seems like there is a conflict. rebase against dev10 again

File: include/seahorn/Expr/ExprOpSort.hh [hunk 1]
```diff
@@ -56,6 +56,9 @@ inline Expr arrayValTy(Expr a) { return a->right(); }
 
 inline Expr structTy(Expr ty) { return mk<STRUCT_TY>(ty); }
 inline Expr structTy(Expr ty1, Expr ty2) { return mk<STRUCT_TY>(ty1, ty2); }
+inline Expr structTy(Expr ty1, Expr ty2, Expr ty3) {
```
Comment (agurfinkel) [inline include/seahorn/Expr/ExprOpSort.hh:59]: not opsem module

File: include/seahorn/Expr/Smt/Z3.hh [hunk 2]
```diff
@@ -377,6 +377,13 @@ public:
 
   Expr eval(Expr e, bool completion = false) {
     assert(model);
```
Comment (agurfinkel) [inline include/seahorn/Expr/Smt/Z3.hh:379]: not opsem module

File: lib/seahorn/BvOpSem2.cc [hunk 3]
```diff
@@ -500,7 +500,9 @@ class OpSemVisitor : public InstVisitor<OpSemVisitor>, OpSemVisitorBase {
                              f->getName().startswith("verifier.nondet") ||
                              f->getName().startswith("__VERIFIER_nondet")))
         visitNondetCall(CS);
-      else
+      else if (f->getName().startswith("__sea_")) {
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2.cc:503]: use longer prefix to avoid capturing unintended functions

File: lib/seahorn/BvOpSem2MemRepr.cc [hunk 4]
```diff
@@ -52,6 +52,9 @@ Expr OpSemMemArrayRepr::MemCpy(Expr dPtr, Expr sPtr, unsigned len,
       Expr val = op::array::select(srcMem, sIdx);
       res = op::array::store(res, dIdx, val);
     }
+  } else {
+    ERR << "Word size and pointer are not aligned!";
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2MemRepr.cc:57]: should be under LOG

File: test/CMakeLists.txt [hunk 5]
```diff
@@ -149,13 +149,23 @@ add_lit_testsuite(test-crab "Regression test for Crab"
...
 
 add_lit_testsuite(test-cex "Regression test for executable counterexamples"
-  -v
-  -j1
-  ${CMAKE_CURRENT_SOURCE_DIR}/cex
-  ARGS
-  --path=${CMAKE_INSTALL_PREFIX}/bin
-  DEPENDS seahorn
-  )
+        -v
```
Comment (agurfinkel) [inline test/CMakeLists.txt:152]: why formatting change?

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -149,13 +149,23 @@ add_lit_testsuite(test-crab "Regression test for Crab"
...
+        -j1
+        ${CMAKE_CURRENT_SOURCE_DIR}/cex
+        ARGS
+        --path=${CMAKE_INSTALL_PREFIX}/bin
+        DEPENDS seahorn
+        )
+
+# Add both and concrete fatptr tests in one target
+add_lit_testsuite(test-fatptr "Regression test for executable counterexamples"
```
Comment (agurfinkel) [inline test/CMakeLists.txt:161]: wrong text in test description

dont need -j 1

File: test/fat_ptr_symbolic/lit.cfg [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+lit_config.note('Repository root is {}'.format(repositoryRoot))
+
+seahorn_cmd = getSeaHorn()
+if not isexec(seahorn_cmd):
+    lit_config.fatal('Could not find the Seahorn executable')
+
+config.substitutions.append(('%seahorn', seahorn_cmd))
+seabmc_fatptr = [seahorn_cmd, '--horn-bv2-lambdas', '--horn-bv2-ptr-size=8',
+                 '--horn-bv2-fatmem',
+                 '--horn-bmc-engine=mono', '--horn-sea-dsa=true',
```
Comment (agurfinkel) [inline test/fat_ptr_symbolic/lit.cfg]: horn-sea-dsa option is dead

File: lib/seahorn/BvOpSem2RawMemMgr.cc [hunk 8]
```diff
@@ -352,10 +353,14 @@ PtrTy RawMemManager::ptrAdd(PtrTy ptr, Expr offset) const {
...
 /// \brief Stores an integer into memory
-///
+/// \param[in] ptr is the address at which _val will be stored
+/// \param[in] _val is the value being written
+/// \param[in] mem is the memory bank/register being written to
+/// \param[in] byteSz is the size of _val in bytes (should be =< word size)
+/// \param[in] align is the known alignment of the ptr
 /// Returns an expression describing the state of memory in \c memReadReg
 /// after the store
-/// \sa loadIntFromMem
```
Comment (agurfinkel) [inline lib/seahorn/BvOpSem2RawMemMgr.cc:358]: typo. should be loadIntFromMeme

File: test/CMakeLists.txt [hunk 9]
```diff
@@ -14,8 +14,8 @@ add_lit_testsuite(test-all "Run all the tests"
 
 add_lit_testsuite(test-fat-ptr "Run fat ptr tests"
   -v
-  ${CMAKE_CURRENT_SOURCE_DIR}/fat_ptr
-  ARGS
+        ${CMAKE_CURRENT_SOURCE_DIR}/fat_ptr_symbolic
+        ARGS
```
Comment (agurfinkel) [inline test/CMakeLists.txt:18]: formatting. 

this replaced fat ptr tests but does not update text description. what about old tests this replaced?

File: test/CMakeLists.txt (hunk 6 above)
Comment (priyasiddharth) [reply test/CMakeLists.txt:161]: ack

File: test/CMakeLists.txt (hunk 5 above)
Comment (priyasiddharth) [reply test/CMakeLists.txt:152]: reverted

File: lib/seahorn/BvOpSem2MemRepr.cc (hunk 4 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2MemRepr.cc:57]: Done

File: lib/seahorn/BvOpSem2.cc (hunk 3 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2.cc:503]: added regex to be more specific.

File: test/fat_ptr_symbolic/lit.cfg (hunk 7 above)
Comment (priyasiddharth) [reply test/fat_ptr_symbolic/lit.cfg]: Ack

File: include/seahorn/Expr/ExprOpSort.hh (hunk 1 above)
Comment (priyasiddharth) [reply include/seahorn/Expr/ExprOpSort.hh:59]: fixed

File: include/seahorn/Expr/Smt/Z3.hh (hunk 2 above)
Comment (priyasiddharth) [reply include/seahorn/Expr/Smt/Z3.hh:379]: fixed

File: lib/seahorn/BvOpSem2RawMemMgr.cc (hunk 8 above)
Comment (priyasiddharth) [reply lib/seahorn/BvOpSem2RawMemMgr.cc:358]: The original comment is wrong. It says loadIntFromMem for storeIntToMem

File: test/CMakeLists.txt (hunk 9 above)
Comment (priyasiddharth) [reply test/CMakeLists.txt:18]: Thanks for catching this! Added new fatptr-symbolic target. Fixed formatting.

**Comment by agurfinkel** [comment]
@priyasiddharth let me know whether this is ready
(you might want to fix commit messages, but this is not crucial)

**Comment by priyasiddharth** [comment]
@agurfinkel This is ready from my end. Not sure what fixing is needed in commit messages. Conventional commits is still not setup at my end.

**Comment by agurfinkel** [comment]
I'll merge.

I've started following conventional commits. There is no automated checker (waiting for you to set it up). One of your commit messages is too long and does not appear well on github. These are all minor things, so I merge and we move on.

**Pull request merged**